// @atproto/api/dist/client/types/app/bsky/actor/defs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateLabelerPrefItem = exports.isLabelerPrefItem = exports.validateLabelersPref = exports.isLabelersPref = exports.validateHiddenPostsPref = exports.isHiddenPostsPref = exports.validateMutedWordsPref = exports.isMutedWordsPref = exports.validateMutedWord = exports.isMutedWord = exports.validateInterestsPref = exports.isInterestsPref = exports.validateThreadViewPref = exports.isThreadViewPref = exports.validateFeedViewPref = exports.isFeedViewPref = exports.validatePersonalDetailsPref = exports.isPersonalDetailsPref = exports.validateSavedFeedsPref = exports.isSavedFeedsPref = exports.validateContentLabelPref = exports.isContentLabelPref = exports.validateAdultContentPref = exports.isAdultContentPref = exports.validateViewerState = exports.isViewerState = exports.validateProfileAssociated = exports.isProfileAssociated = exports.validateProfileViewDetailed = exports.isProfileViewDetailed = exports.validateProfileView = exports.isProfileView = exports.validateProfileViewBasic = exports.isProfileViewBasic = void 0;
const util_1 = require("../../../../util");
const lexicons_1 = require("../../../../lexicons");
function isProfileViewBasic(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#profileViewBasic');
}
exports.isProfileViewBasic = isProfileViewBasic;
function validateProfileViewBasic(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#profileViewBasic', v);
}
exports.validateProfileViewBasic = validateProfileViewBasic;
function isProfileView(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#profileView');
}
exports.isProfileView = isProfileView;
function validateProfileView(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#profileView', v);
}
exports.validateProfileView = validateProfileView;
function isProfileViewDetailed(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#profileViewDetailed');
}
exports.isProfileViewDetailed = isProfileViewDetailed;
function validateProfileViewDetailed(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#profileViewDetailed', v);
}
exports.validateProfileViewDetailed = validateProfileViewDetailed;
function isProfileAssociated(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#profileAssociated');
}
exports.isProfileAssociated = isProfileAssociated;
function validateProfileAssociated(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#profileAssociated', v);
}
exports.validateProfileAssociated = validateProfileAssociated;
function isViewerState(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#viewerState');
}
exports.isViewerState = isViewerState;
function validateViewerState(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#viewerState', v);
}
exports.validateViewerState = validateViewerState;
function isAdultContentPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#adultContentPref');
}
exports.isAdultContentPref = isAdultContentPref;
function validateAdultContentPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#adultContentPref', v);
}
exports.validateAdultContentPref = validateAdultContentPref;
function isContentLabelPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#contentLabelPref');
}
exports.isContentLabelPref = isContentLabelPref;
function validateContentLabelPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#contentLabelPref', v);
}
exports.validateContentLabelPref = validateContentLabelPref;
function isSavedFeedsPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#savedFeedsPref');
}
exports.isSavedFeedsPref = isSavedFeedsPref;
function validateSavedFeedsPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#savedFeedsPref', v);
}
exports.validateSavedFeedsPref = validateSavedFeedsPref;
function isPersonalDetailsPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#personalDetailsPref');
}
exports.isPersonalDetailsPref = isPersonalDetailsPref;
function validatePersonalDetailsPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#personalDetailsPref', v);
}
exports.validatePersonalDetailsPref = validatePersonalDetailsPref;
function isFeedViewPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#feedViewPref');
}
exports.isFeedViewPref = isFeedViewPref;
function validateFeedViewPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#feedViewPref', v);
}
exports.validateFeedViewPref = validateFeedViewPref;
function isThreadViewPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#threadViewPref');
}
exports.isThreadViewPref = isThreadViewPref;
function validateThreadViewPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#threadViewPref', v);
}
exports.validateThreadViewPref = validateThreadViewPref;
function isInterestsPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#interestsPref');
}
exports.isInterestsPref = isInterestsPref;
function validateInterestsPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#interestsPref', v);
}
exports.validateInterestsPref = validateInterestsPref;
function isMutedWord(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#mutedWord');
}
exports.isMutedWord = isMutedWord;
function validateMutedWord(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#mutedWord', v);
}
exports.validateMutedWord = validateMutedWord;
function isMutedWordsPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#mutedWordsPref');
}
exports.isMutedWordsPref = isMutedWordsPref;
function validateMutedWordsPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#mutedWordsPref', v);
}
exports.validateMutedWordsPref = validateMutedWordsPref;
function isHiddenPostsPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#hiddenPostsPref');
}
exports.isHiddenPostsPref = isHiddenPostsPref;
function validateHiddenPostsPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#hiddenPostsPref', v);
}
exports.validateHiddenPostsPref = validateHiddenPostsPref;
function isLabelersPref(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#labelersPref');
}
exports.isLabelersPref = isLabelersPref;
function validateLabelersPref(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#labelersPref', v);
}
exports.validateLabelersPref = validateLabelersPref;
function isLabelerPrefItem(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'app.bsky.actor.defs#labelerPrefItem');
}
exports.isLabelerPrefItem = isLabelerPrefItem;
function validateLabelerPrefItem(v) {
    return lexicons_1.lexicons.validate('app.bsky.actor.defs#labelerPrefItem', v);
}
exports.validateLabelerPrefItem = validateLabelerPrefItem;
