// @atproto/api/dist/client/types/com/atproto/admin/defs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateRepoBlobRef = exports.isRepoBlobRef = exports.validateRepoRef = exports.isRepoRef = exports.validateAccountView = exports.isAccountView = exports.validateStatusAttr = exports.isStatusAttr = void 0;
const util_1 = require("../../../../util");
const lexicons_1 = require("../../../../lexicons");
function isStatusAttr(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'com.atproto.admin.defs#statusAttr');
}
exports.isStatusAttr = isStatusAttr;
function validateStatusAttr(v) {
    return lexicons_1.lexicons.validate('com.atproto.admin.defs#statusAttr', v);
}
exports.validateStatusAttr = validateStatusAttr;
function isAccountView(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'com.atproto.admin.defs#accountView');
}
exports.isAccountView = isAccountView;
function validateAccountView(v) {
    return lexicons_1.lexicons.validate('com.atproto.admin.defs#accountView', v);
}
exports.validateAccountView = validateAccountView;
function isRepoRef(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'com.atproto.admin.defs#repoRef');
}
exports.isRepoRef = isRepoRef;
function validateRepoRef(v) {
    return lexicons_1.lexicons.validate('com.atproto.admin.defs#repoRef', v);
}
exports.validateRepoRef = validateRepoRef;
function isRepoBlobRef(v) {
    return ((0, util_1.isObj)(v) &&
        (0, util_1.hasProp)(v, '$type') &&
        v.$type === 'com.atproto.admin.defs#repoBlobRef');
}
exports.isRepoBlobRef = isRepoBlobRef;
function validateRepoBlobRef(v) {
    return lexicons_1.lexicons.validate('com.atproto.admin.defs#repoBlobRef', v);
}
exports.validateRepoBlobRef = validateRepoBlobRef;
